package/libpiksi/libpiksi/src/endpoint.c

```diff
@@ -100,196 +152,114 @@ pk_endpoint_type pk_endpoint_type_get(pk_endpoint_t *pk_ept)
 
-  ssize_t result = fill_buffer_from_zmsg(msg, buffer, count);
-
-  zmsg_destroy(&msg);
-  assert(msg == NULL);
-
-  return result;
-}
-
-/**
- * @brief dup_zmsg_to_buffer - helper to collapse zmsg frames
- * @note  caller must free the returned buffer if call is successful
- * @param msg: zmq message to copy
- * @param buffer_loc: double pointer to place alloc'd buffer
- * @param count_loc: pointer to place the length of the returned data
- * @return 0 on success, -1 on failure
- */
-static int dup_zmsg_to_buffer(zmsg_t *msg, u8 **buffer_loc, size_t *count_loc)
-{
-  assert(msg != NULL);
-  assert(buffer_loc != NULL);
-  assert(count_loc != NULL);
-
-  size_t buffer_size = zmsg_content_size(msg);
-  u8 *buffer = (u8 *)malloc(buffer_size);
-  if (buffer == NULL) {
-    piksi_log(LOG_ERR, "Failed to allocate buffer for zmsg dup");
-    return -1;
-  }
-
-  ssize_t buffer_index = fill_buffer_from_zmsg(msg, buffer, buffer_size);
-  if (buffer_index == -1) {
-    piksi_log(LOG_ERR, "failed to copy zmsg into dup buffer");
-    return -1;
-  }
-
-  if (buffer_index != buffer_size) {
-    piksi_log(LOG_WARNING, "dup zmsg - data copied not equal to buffer_size");
-  }
-  *buffer_loc = buffer;
-  *count_loc = buffer_index;
-  return 0;
+  return length;
 }
 
 int pk_endpoint_receive(pk_endpoint_t *pk_ept, pk_endpoint_receive_cb rx_cb, void *context)
 {
   assert(pk_ept != NULL);
-  assert(pk_ept->type != PK_ENDPOINT_PUB);
+  assert(pk_ept->type != PK_ENDPOINT_PUB || pk_ept->type != PK_ENDPOINT_PUB_SERVER);
   assert(rx_cb != NULL);
 
-  while (ZMQ_POLLIN & zsock_events(pk_ept->zsock)) {
-    zmsg_t * msg = pk_endpoint_receive_zmsg(pk_ept);
-    if (msg == NULL) {
-      return -1;
-    }
-
-    if (zmsg_size(msg) != 1) {
-      piksi_log(LOG_WARNING, "Piksi endpoint received non-conforming message!");
-    }
+  u8 *buffer = NULL;
+  size_t length = NN_MSG;
+  if (pk_endpoint_receive_nn_msg(pk_ept, (void *)&buffer, &length) != 0) {
+    piksi_log(LOG_ERR, "failed to receive nn_msg");
+    return -1;
+  }
 
-    u8 *dup_buffer = NULL;
-    size_t dup_length = 0;
-    if (dup_zmsg_to_buffer(msg, &dup_buffer, &dup_length) == 0) {
-      rx_cb(dup_buffer, dup_length, context);
-      free(dup_buffer);
-    }
+  rx_cb(buffer, length, context);
+  nn_freemsg(buffer);
 
-    zmsg_destroy(&msg);
-  }
   return 0;
 }
 
 int pk_endpoint_send(pk_endpoint_t *pk_ept, const u8 *data, const size_t length)
 {
   assert(pk_ept != NULL);
-  assert(pk_ept->type != PK_ENDPOINT_SUB);
-
-  zmsg_t *msg = zmsg_new();
-  if (msg == NULL) {
-    piksi_log(LOG_ERR, "error in zmsg_new()");
-    return -1;
-  }
-
-  if (zmsg_addmem(msg, data, length) != 0) {
-    piksi_log(LOG_ERR, "error in zmsg_addmem()");
-    zmsg_destroy(&msg);
-    return -1;
-  }
+  assert(pk_ept->type != PK_ENDPOINT_SUB || pk_ept->type != PK_ENDPOINT_SUB_SERVER);
 
   while (1) {
-    int ret = zmsg_send(&msg, pk_ept->zsock);
-    if (ret == 0) {
+    int written = nn_send(pk_ept->nn_sock, data, length, 0);
```

Comment:
Might be nice to expose the nn_allocmg  API through our wrapper so we can avoid some buffer copies.